### ISSUE TITLE
Type Errors: I can't build the project or push it to Heroku for example

### ISSUE THREAD

**Issue by jeremienegie** [telegraf/telegraf#1490]
So I'm using typescript and express and to build a bot but I can't push my project to Heroku because when I run the build this is the error that I'm faced with and I went into the files and it's really just simple errors

```
node_modules/telegraf/src/core/network/client.ts:347:11 - error TS2578: Unused '@ts-expect-error' directive.

347           // @ts-expect-error cannot assign to Record<string, unknown>
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/telegraf/src/markup.ts:124:28 - error TS2345: Argument of type '((string & { hide?: boolean | undefined; }) | (CommonButton & { hide?: boolean | undefined; }))[]' is not assignable to parameter of type 'B[]'.
  Type '(string & { hide?: boolean | undefined; }) | (CommonButton & { hide?: boolean | undefined; })' is not assignable to type 'B'.
    '(string & { hide?: boolean | undefined; }) | (CommonButton & { hide?: boolean | undefined; })' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'HideableKBtn | HideableIKBtn'.
      Type 'string & { hide?: boolean | undefined; }' is not assignable to type 'B'.
        'string & { hide?: boolean | undefined; }' is assignable to the constraint of type 'B', but 'B' could be instantiated with a different subtype of constraint 'HideableKBtn | HideableIKBtn'.

124     if (wrapFn(btn, index, currentRow) && currentRow.length > 0) {
                               ~~~~~~~~~~

node_modules/telegraf/src/markup.ts:125:19 - error TS2345: Argument of type '((string & { hide?: boolean | undefined; }) | (CommonButton & { hide?: boolean | undefined; }))[]' is not assignable to parameter of type 'B[]'.

125       result.push(currentRow)
                      ~~~~~~~~~~

node_modules/telegraf/src/markup.ts:132:17 - error TS2345: Argument of type '((string & { hide?: boolean | undefined; }) | (CommonButton & { hide?: boolean | undefined; }))[]' is not assignable to parameter of type 'B[]'.

132     result.push(currentRow)
                    ~~~~~~~~~~


Found 4 errors.
```

**Comment by wojpawlik** [comment]
You seem to be using TypeScript 4.3, while Telegraf is sticking with 4.2 until 4.4. Workaround: `skipLibCheck: true`.

**Comment by wojpawlik** [comment]
Actually, it's weird that your setup is typechecking Telegraf rather than relying on `.d.ts`

**Comment by jeremienegie** [comment]
> Actually, it's weird that your setup is typechecking Telegraf rather than relying on `.d.ts`

I'm also confused and I have set skipLibCheck: true

here is my tsconfig file

```
{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "lib": ["esnext"],
    "allowJs": true,
    "outDir": "build",
    "rootDir": "src",
    "strict": true,
    "noImplicitAny": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "types": ["node", "@types/jest"],
    "typeRoots": ["./node_modules/@types"],
    "skipLibCheck": true
  },
  "include": ["src/**/*", "global.d.ts"],
  "exclude": ["src/**/*.spec.ts"]
}
```

**Comment by jeremienegie** [comment]
> You seem to be using TypeScript 4.3, while Telegraf is sticking with 4.2 until 4.4. Workaround: `skipLibCheck: true`.

Here is my package.json

```
{
  "name": "Telegram Bot",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "build": "rimraf ./build && tsc",
    "start:dev": "nodemon",
    "start": "yarn run build && node build/index.js",
    "lint": "eslint . --ext .ts",
    "prettier-format": "run-script-os",
    "prettier-format:win32": "prettier --config .prettierrc \"./src/**/*.ts\" --write",
    "prettier-format:darwin:linux": "prettier --config .prettierrc 'src/**/*.ts' --write",
    "prettier-format:default": "prettier --config .prettierrc 'src/**/*.ts' --write",
    "prettier-watch": "run-script-os",
    "prettier-watch:win32": "onchange \"src/**/*.ts\" -- prettier --write {{changed}}",
    "prettier-watch:darwin:linux": "onchange 'src/**/*.ts' -- prettier --write {{changed}}",
    "prettier-watch:default": "onchange 'src/**/*.ts' -- prettier --write {{changed}}",
    "test": "jest",
    "test:dev": "jest --watchAll",
    "prepare": "husky install"
  },
  "husky": {
    "hooks": {
      "pre-commit": "yarn run test && yarn run prettier-format && yarn run lint"
    }
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/jest": "^26.0.24",
    "@types/node": "^16.4.1",
    "@types/node-fetch": "^2.5.11",
    "@typescript-eslint/eslint-plugin": "^4.28.4",
    "@typescript-eslint/parser": "^4.28.4",
    "eslint": "^7.31.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^3.1.2",
    "husky": "^7.0.0",
    "nodemon": "^2.0.12",
    "onchange": "^7.1.0",
    "prettier": "^2.3.2",
    "rimraf": "^3.0.0",
    "run-script-os": "^1.1.1",
    "ts-node": "^10.1.0",
    "typescript": "^4.2.0"
  },
  "dependencies": {
    "@types/coinbase-commerce-node": "^1.0.6",
    "@vonage/server-sdk": "^2.10.10",
    "axios": "^0.21.1",
    "coinbase-commerce-node": "^1.0.4",
    "contentful-management": "^7.30.0",
    "dotenv": "^10.0.0",
    "eslint-plugin-jest": "^24.1.0",
    "express": "^4.17.1",
    "jest": "^27.0.6",
    "jssip": "^3.8.0",
    "jssip-node-websocket": "^3.0.0",
    "moment": "^2.29.1",
    "telegraf": "^4.4.1",
    "ts-jest": "^27.0.4"
  }
}

```

**Comment by 0wx** [comment]
> > You seem to be using TypeScript 4.3, while Telegraf is sticking with 4.2 until 4.4. Workaround: `skipLibCheck: true`.
> 
> Here is my package.json
> 
> ```
> {
>   "name": "Telegram Bot",
>   "version": "1.0.0",
>   "description": "",
>   "main": "index.js",
>   "scripts": {
>     "build": "rimraf ./build && tsc",
>     "start:dev": "nodemon",
>     "start": "yarn run build && node build/index.js",
>     "lint": "eslint . --ext .ts",
>     "prettier-format": "run-script-os",
>     "prettier-format:win32": "prettier --config .prettierrc \"./src/**/*.ts\" --write",
>     "prettier-format:darwin:linux": "prettier --config .prettierrc 'src/**/*.ts' --write",
>     "prettier-format:default": "prettier --config .prettierrc 'src/**/*.ts' --write",
>     "prettier-watch": "run-script-os",
>     "prettier-watch:win32": "onchange \"src/**/*.ts\" -- prettier --write {{changed}}",
>     "prettier-watch:darwin:linux": "onchange 'src/**/*.ts' -- prettier --write {{changed}}",
>     "prettier-watch:default": "onchange 'src/**/*.ts' -- prettier --write {{changed}}",
>     "test": "jest",
>     "test:dev": "jest --watchAll",
>     "prepare": "husky install"
>   },
>   "husky": {
>     "hooks": {
>       "pre-commit": "yarn run test && yarn run prettier-format && yarn run lint"
>     }
>   },
>   "keywords": [],
>   "author": "",
>   "license": "ISC",
>   "devDependencies": {
>     "@types/express": "^4.17.13",
>     "@types/jest": "^26.0.24",
>     "@types/node": "^16.4.1",
>     "@types/node-fetch": "^2.5.11",
>     "@typescript-eslint/eslint-plugin": "^4.28.4",
>     "@typescript-eslint/parser": "^4.28.4",
>     "eslint": "^7.31.0",
>     "eslint-config-prettier": "^8.3.0",
>     "eslint-plugin-prettier": "^3.1.2",
>     "husky": "^7.0.0",
>     "nodemon": "^2.0.12",
>     "onchange": "^7.1.0",
>     "prettier": "^2.3.2",
>     "rimraf": "^3.0.0",
>     "run-script-os": "^1.1.1",
>     "ts-node": "^10.1.0",
>     "typescript": "^4.2.0"
>   },
>   "dependencies": {
>     "@types/coinbase-commerce-node": "^1.0.6",
>     "@vonage/server-sdk": "^2.10.10",
>     "axios": "^0.21.1",
>     "coinbase-commerce-node": "^1.0.4",
>     "contentful-management": "^7.30.0",
>     "dotenv": "^10.0.0",
>     "eslint-plugin-jest": "^24.1.0",
>     "express": "^4.17.1",
>     "jest": "^27.0.6",
>     "jssip": "^3.8.0",
>     "jssip-node-websocket": "^3.0.0",
>     "moment": "^2.29.1",
>     "telegraf": "^4.4.1",
>     "ts-jest": "^27.0.4"
>   }
> }
> ```

Try add `"postinstall": "tsc"` in your scripts and remove the `yarn run build` from the `start` scripts.

**Comment by jeremienegie** [comment]
> > > You seem to be using TypeScript 4.3, while Telegraf is sticking with 4.2 until 4.4. Workaround: `skipLibCheck: true`.
> > 
> > 
> > Here is my package.json
> > ```
> > {
> >   "name": "Telegram Bot",
> >   "version": "1.0.0",
> >   "description": "",
> >   "main": "index.js",
> >   "scripts": {
> >     "build": "rimraf ./build && tsc",
> >     "start:dev": "nodemon",
> >     "start": "yarn run build && node build/index.js",
> >     "lint": "eslint . --ext .ts",
> >     "prettier-format": "run-script-os",
> >     "prettier-format:win32": "prettier --config .prettierrc \"./src/**/*.ts\" --write",
> >     "prettier-format:darwin:linux": "prettier --config .prettierrc 'src/**/*.ts' --write",
> >     "prettier-format:default": "prettier --config .prettierrc 'src/**/*.ts' --write",
> >     "prettier-watch": "run-script-os",
> >     "prettier-watch:win32": "onchange \"src/**/*.ts\" -- prettier --write {{changed}}",
> >     "prettier-watch:darwin:linux": "onchange 'src/**/*.ts' -- prettier --write {{changed}}",
> >     "prettier-watch:default": "onchange 'src/**/*.ts' -- prettier --write {{changed}}",
> >     "test": "jest",
> >     "test:dev": "jest --watchAll",
> >     "prepare": "husky install"
> >   },
> >   "husky": {
> >     "hooks": {
> >       "pre-commit": "yarn run test && yarn run prettier-format && yarn run lint"
> >     }
> >   },
> >   "keywords": [],
> >   "author": "",
> >   "license": "ISC",
> >   "devDependencies": {
> >     "@types/express": "^4.17.13",
> >     "@types/jest": "^26.0.24",
> >     "@types/node": "^16.4.1",
> >     "@types/node-fetch": "^2.5.11",
> >     "@typescript-eslint/eslint-plugin": "^4.28.4",
> >     "@typescript-eslint/parser": "^4.28.4",
> >     "eslint": "^7.31.0",
> >     "eslint-config-prettier": "^8.3.0",
> >     "eslint-plugin-prettier": "^3.1.2",
> >     "husky": "^7.0.0",
> >     "nodemon": "^2.0.12",
> >     "onchange": "^7.1.0",
> >     "prettier": "^2.3.2",
> >     "rimraf": "^3.0.0",
> >     "run-script-os": "^1.1.1",
> >     "ts-node": "^10.1.0",
> >     "typescript": "^4.2.0"
> >   },
> >   "dependencies": {
> >     "@types/coinbase-commerce-node": "^1.0.6",
> >     "@vonage/server-sdk": "^2.10.10",
> >     "axios": "^0.21.1",
> >     "coinbase-commerce-node": "^1.0.4",
> >     "contentful-management": "^7.30.0",
> >     "dotenv": "^10.0.0",
> >     "eslint-plugin-jest": "^24.1.0",
> >     "express": "^4.17.1",
> >     "jest": "^27.0.6",
> >     "jssip": "^3.8.0",
> >     "jssip-node-websocket": "^3.0.0",
> >     "moment": "^2.29.1",
> >     "telegraf": "^4.4.1",
> >     "ts-jest": "^27.0.4"
> >   }
> > }
> > ```
> 
> Try add `"postinstall": "tsc"` in your scripts and remove the `yarn run build` from the `start` scripts.

I have tried that but still getting the same errors

**Comment by jeremienegie** [comment]
I just tried today and it's worked but I have no idea why

**Comment by abhint** [comment]
This is just an example. You can replace `app.js` and `app.ts` with your file name.

```bash
package.json
```

```js
  "scripts": {
    "start": "node app.js",
    "test": "nodemon app.ts",
    "build": "tsc",
    "watch-node": "nodemon app.js",
    "watch-ts": "tsc -w",
    "deploy": "git add . && git commit -m Heroku && git push heroku master"
  },
```

**Comment by BALEHOK** [comment]
I had similar issues with building a project with `telegraf`. I figured the problem in my case was that I imported a couple of types from the `src` folder, like `import SceneContextScene from 'telegraf/src/scenes/context';`. After I replaced `src` by `typings` the errors were gone: `import SceneContextScene from 'telegraf/typings/scenes/context';`.
Bottom line: don't import from `telegraf/src` folder.
